src/testing_tracer.rs

```diff
@@ -0,0 +1,209 @@
+            self.exception_message = message[2..].to_owned();
+        } else {
+            self.reset_exception();
+        }
+    }
+
+    fn update_expecting_register_value(&mut self, message: &str) {
+        if message.starts_with(PRINT_REG_PREFIX) {
+            self.expecting_register_value = true;
+        } else {
+            self.expecting_register_value = false;
+        }
+    }
+}
+
+impl Tracer for TestingTracer {
+    type SupportedMemory = SimpleMemory;
+    const CALL_BEFORE_EXECUTION: bool = true;
+    const CALL_AFTER_DECODING: bool = true;
+
+    #[inline]
+    fn before_decoding(&mut self, _state: VmLocalStateData<'_>, _memory: &Self::SupportedMemory) {}
+
+    fn after_decoding(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: AfterDecodingData,
+        _memory: &Self::SupportedMemory,
+    ) {
+        if _data.error_flags_accumulated.is_empty() {
+            return;
+        }
+
+        self.has_exception = true;
+
+        if _data
+            .error_flags_accumulated
+            .contains(ErrorFlags::NOT_ENOUGH_ERGS)
+        {
+            self.exception_message = "NOT_ENOUGH_ERGS".to_owned();
+        }
+        if _data
+            .error_flags_accumulated
+            .contains(ErrorFlags::INVALID_OPCODE)
+        {
+            self.exception_message = "INVALID OPCODE".to_owned();
+        }
+        if _data
+            .error_flags_accumulated
+            .contains(ErrorFlags::PRIVILAGED_ACCESS_NOT_FROM_KERNEL)
+        {
+            self.exception_message = "PRIVILAGED_ACCESS_NOT_FROM_KERNEL".to_owned();
+        }
+        if _data
+            .error_flags_accumulated
+            .contains(ErrorFlags::WRITE_IN_STATIC_CONTEXT)
+        {
+            self.exception_message = "WRITE_IN_STATIC_CONTEXT".to_owned();
+        }
+        if _data
+            .error_flags_accumulated
+            .contains(ErrorFlags::CALLSTACK_IS_FULL)
+        {
+            self.exception_message = "CALLSTACK_IS_FULL".to_owned();
+        }
+    }
+
+    fn before_execution(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: BeforeExecutionData,
+        _memory: &Self::SupportedMemory,
+    ) {
+        /*
+        println!("{}", _data.opcode);
+        println!("New pc: {}", _data.new_pc);
+        println!("Val: {}", _data.src0_value.value);
+        println!("");
+        */
+        let inner_opcode = _data.opcode.inner.variant.opcode;
+
+        match inner_opcode {
+            Opcode::Ret(RetOpcode::Panic) => {
+                if self.exception_message != "" {
+                    self.has_exception = true;
+                }
+            }
+            Opcode::Add(AddOpcode::Add) => {
+                // `add x r0 r0` is used as "print" statement
+                if _data.opcode.dst0_reg_idx == 0 {
+                    if self.expecting_register_value {
+                        println!("{}", _data.src0_value.value);
```

Comment:
remove?

Reply:
This is the task of this command - to print to the console. Changed the structure to make it a bit clearer.